feat(Computability/DFA): Closure of regular languages under some set operations

maemre opened the pull request:
This shows that regular languages are closed under complement and intersection by constructing DFAs for them.

---

Closure under all other operations will be proved when someone adds the proof for DFA<->regular expression equivalence, so they are not part of this PR.

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

maemre committed:
Closure of regular languages under some set operations.

EtienneC30 commented on Mathlib/Computability/DFA.lean:
```diff
@@ -164,6 +164,82 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+
+theorem cartesian_product_accept_iff
+    (p : Prop → Prop → Prop) (M₁ : DFA α σ) (M₂ : DFA α σ') (s : σ × σ') :
+    s ∈ (cartesian_product p M₁ M₂).accept ↔ p (s.fst ∈ M₁.accept) (s.snd ∈ M₂.accept) := by
```
You can also write `M₁.cartesian_product p M₂`. This feels a bit weird with the `p` in the middle, so you could move the p after the DFAs in the arguments of the def. This is just a suggestion, feel free to keep it as it is if you prefer.

maemre replied:
That syntax is much clearer, thanks for the suggestion!

EtienneC30 reviewed (commented):
Thank you for this! Here are a few comments, also with suggestions for better readability.

maemre committed:
Fixed style issues from code review.

maemre committed:
Fixed struct construction style to match mathlib style.

maemre commented on Mathlib/Computability/DFA.lean:
```diff
@@ -164,6 +164,82 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+instance : HasCompl (DFA α σ) := ⟨compl⟩
+
+theorem invert_accept_iff (s : σ) : s ∈ Mᶜ.accept ↔ s ∉ M.accept := by
+  simp [HasCompl.compl, compl]
```
I don't exactly understand this suggestion.

Do you want to add another lemma like the following before this then just use `simp` here?

```lean
theorem compl_accept_eq_accept_compl (s : σ) : Mᶜ.accept = M.acceptᶜ := by ...
```

maemre replied:
I added the lemma I mentioned above.

maemre commented on Mathlib/Computability/DFA.lean:
```diff
@@ -164,6 +164,82 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+
+ This is a generalization of the intersection of two DFAs to arbitrary binary set operations.
+-/
+def cartesian_product (p : Prop → Prop → Prop) (M₁ : DFA α σ) (M₂ : DFA α σ') : DFA α (σ × σ') := {
```
changed the name to `product`

maemre commented on Mathlib/Computability/DFA.lean:
```diff
@@ -164,6 +164,82 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+    x ∈ (cartesian_product p M₁ M₂).accepts ↔ p (x ∈ M₁.accepts) (x ∈ M₂.accepts) := by
+  simp [DFA.accepts]
+  simp [DFA.acceptsFrom, cartesian_product_eval_from]
+  simp [cartesian_product]
```
I think I applied your suggestion correctly, but I am not sure.

EtienneC30 replied:
Yes that was what I meant. The use of `mem_setOf_eq` was bugging me because it's not something we usually write explicitly, and in fact you can just do this for the proof:
```lean
theorem product_accepts_iff
    (p : Prop → Prop → Prop) (M₁ : DFA α σ) (M₂ : DFA α σ') (x : List α) :
    x ∈ (M₁.product M₂ p).accepts ↔ p (x ∈ M₁.accepts) (x ∈ M₂.accepts) := by
  simp only [accepts, acceptsFrom, product_evalFrom]
  rfl
```

maemre replied:
Thank you, I just changed the proof to this.

maemre commented on Mathlib/Computability/DFA.lean:
```diff
@@ -164,6 +164,82 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+    (p : Prop → Prop → Prop) (M₁ : DFA α σ) (M₂ : DFA α σ') (x : List α) :
+    ∀ s : σ × σ',
+      (cartesian_product p M₁ M₂).evalFrom s x = ⟨M₁.evalFrom s.1 x, M₂.evalFrom s.2 x⟩ := by
+  dsimp [DFA.evalFrom, cartesian_product]
```
thanks, fixed this in other places too.

maemre commented on Mathlib/Computability/DFA.lean:
```diff
@@ -164,6 +164,82 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+theorem invert_accept_iff (s : σ) : s ∈ Mᶜ.accept ↔ s ∉ M.accept := by
+  simp [HasCompl.compl, compl]
+
+theorem invert_accepts_iff (x : List α) : x ∈ Mᶜ.accepts ↔ x ∉ M.accepts := by
```
This is also how the intersection theorem is stated in #22298. I think that is useful, and I will add those lemmas too.

maemre commented on Mathlib/Computability/DFA.lean:
```diff
@@ -164,6 +164,82 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+
+instance : HasCompl (DFA α σ) := ⟨compl⟩
+
+theorem invert_accept_iff (s : σ) : s ∈ Mᶜ.accept ↔ s ∉ M.accept := by
```
Changed it to `compl`.

maemre commented on Mathlib/Computability/DFA.lean:
```diff
@@ -164,6 +164,82 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+theorem cartesian_product_eval_from
+    (p : Prop → Prop → Prop) (M₁ : DFA α σ) (M₂ : DFA α σ') (x : List α) :
+    ∀ s : σ × σ',
+      (cartesian_product p M₁ M₂).evalFrom s x = ⟨M₁.evalFrom s.1 x, M₂.evalFrom s.2 x⟩ := by
```
Thank you, I didn't know about the `revert` tactic. I changed the statement.

maemre committed:
use induction rather than induction'

maemre committed:
add a proof of closure under union

maemre committed:
fix indentation

maemre committed:
Set notation-based lemmas for DFA.compl

maemre committed:
Set notation-based lemma for the accepting states for DFA intersection

maemre commented on Mathlib/Computability/DFA.lean:
```diff
@@ -164,6 +164,82 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+    ∀ s : σ × σ',
+      (cartesian_product p M₁ M₂).evalFrom s x = ⟨M₁.evalFrom s.1 x, M₂.evalFrom s.2 x⟩ := by
+  dsimp [DFA.evalFrom, cartesian_product]
+  induction' x with a x' ih
```
I switched to `induction`. Does the result look like the preferred style in mathlib4?

EtienneC30 replied:
Yes that's it ,thanks!

maemre commented:
@EtienneC30 Adding set notation for the closure properties requires adding `Union` and `Intersection` instances to `Language`, which is not defined in this module (it is defined in `Computability/Language`). Should I add them as part of this PR or create a new PR?

EtienneC30 commented:
You can add them in this PR it's fine. The union of languages is actually already defined there but it is registered with an `Add` instance, so the notation is `+`. There is nothing for the intersection, I don't know if there is a special convention for the intersection of languages which uses something other than `∩`, but I guess if it was not registered it's not the case, so you can go with `∩`.

EtienneC30 commented:
Don't forget to remove the awaiting-author label when you are done.

maemre committed:
Remove unnecessary lemma in the proof of intersect_accept_product_accept

maemre committed:
Add set operation notation to languages

maemre committed:
Set notation proofs for DFA closure properties

maemre committed:
replace rewrites from Set lemmas with just rfl

maemre committed:
Use the existing Add instance for Language rather than adding a Union instance.

maemre commented:
@EtienneC30 I added the `Inter` instance for `Language` and added the set notation-based proofs to `Computability/DFA`. I also removed explicit uses of `Set` methods in the closure proofs as you suggested. I am done with all changes.

EtienneC30 commented:
Can you please merge master as suggested by the bot? I don't why it asks, maybe it's because the branch is a bit old, but I think it's better if you do it.

EtienneC30 commented on Mathlib/Computability/DFA.lean:
```diff
@@ -164,6 +164,116 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+
+theorem compl_accept_iff (s : σ) : s ∈ Mᶜ.accept ↔ s ∉ M.accept := by simp
+
+theorem compl_accepts_iff (x : List α) : x ∈ Mᶜ.accepts ↔ x ∉ M.accepts := by
+  apply compl_accept_iff
```
A proof which is just `by apply ...` can usually be replaced by a term-mode proof, which is nicer:
```suggestion
theorem compl_accepts_iff (x : List α) : x ∈ Mᶜ.accepts ↔ x ∉ M.accepts :=
  compl_accept_iff M _
```

EtienneC30 commented on Mathlib/Computability/DFA.lean:
```diff
@@ -164,6 +164,116 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+
+ There is no instance for this to provide the `M₁ ∩ M₂` syntax, because M₁ and M₂ have
+ different state types. -/
+def intersect (M₁ : DFA α σ) (M₂ : DFA α σ') : DFA α (σ × σ') := product M₁ M₂ And
```
I think this should be called `inter` because that's how we refer to intersection usually. You should also change the name of all the theorems which use `intersect`.
```suggestion
def inter (M₁ : DFA α σ) (M₂ : DFA α σ') : DFA α (σ × σ') := product M₁ M₂ And
```

EtienneC30 commented on Mathlib/Computability/DFA.lean:
```diff
@@ -164,6 +164,116 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+    x ∈ (M₁.intersect M₂).accepts ↔ x ∈ M₁.accepts ∧ x ∈ M₂.accepts := by
+  simp [intersect, product_accepts_iff]
+
+theorem intersect_accept_eq_product_accept (M₁ : DFA α σ) (M₂ : DFA α σ') :
```
The product of sets is referred to as `prod` in names.
```suggestion
theorem inter_accept_eq_prod_accept (M₁ : DFA α σ) (M₂ : DFA α σ') :
```

EtienneC30 replied:
You did not apply this.

maemre replied:
I am sorry for missing this. Just fixed it.

EtienneC30 commented on Mathlib/Computability/DFA.lean:
```diff
@@ -164,6 +164,116 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+  ext ⟨s₁, s₂⟩
+  simp [intersect_accept_iff]
+
+theorem intersect_accepts_intersect (M₁ : DFA α σ) (M₂ : DFA α σ') :
```
I would suggest this name:
```suggestion
theorem inter_accepts_eq_inter(M₁ : DFA α σ) (M₂ : DFA α σ') :
```

maemre replied:
Applied this name.

EtienneC30 reviewed (commented):
This seems fine, just a few remarks about naming.

maemre committed:
Merge branch 'master' into dfa_set_closure

maemre committed:
replace " Mathlib/Computability/DFA.lean

Co-authored-by: EtienneC30 <66847262+EtienneC30@users.noreply.github.com>

maemre committed:
Rename intersect to inter.

maemre commented:
I renamed `intersect` to `inter` and replaced the `apply` with the term-mode proof.

EtienneC30 commented:
Please address all the remarks before removing the awaiting-author label. If you applied a suggestion, you can resolve the conversation, this makes things easier to read. If you disagree with a suggestion, please say it so that we can discuss it.

maemre committed:
Rename product to prod and clean up the name of a lemma.

maemre commented:
I am sorry for missing the last two remarks. I just fixed them and resolved the conversation. I will be careful about the PR workflow going forward.

EtienneC30 commented:
No problem, it takes a bit of time to get used to the review process. This PR is in good shape I think, now we have to wait for a maintainer to take a look at it and merge it. Thank you for your work!
maintainer merge

urkud commented:
Could you please add statements in terms of `Language.IsRegular`?

YaelDillies commented on Mathlib/Computability/DFA.lean:
```diff
@@ -164,6 +164,116 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+variable {σ' : Type v}
+
+/--
+ `M.compl` constructs a DFA for the complement of the language of `M`.
+
+ Use `Mᶜ` rather than directly using this function.
```
```suggestion
`M.compl` constructs a DFA for the complement of the language of `M`.

Use `Mᶜ` rather than directly using this function.
```
Same below

YaelDillies commented on Mathlib/Computability/DFA.lean:
```diff
@@ -164,6 +164,116 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
+@[simp]
+theorem compl_accepts_eq_accepts_compl : Mᶜ.accepts = M.acceptsᶜ := rfl
+
+theorem compl_accept_iff (s : σ) : s ∈ Mᶜ.accept ↔ s ∉ M.accept := by simp
```
```suggestion
theorem mem_accept_compl {s : σ} : s ∈ Mᶜ.accept ↔ s ∉ M.accept := by simp
```
Same below

maemre replied:
I made these changes for the theorems about complements. For other set operations (union and intersection), what would be good choices for making parameters implicit w.r.t. mathlib4 expectations?

For example, currently I have:

```lean
theorem inter_accept_iff (M₁ : DFA α σ) (M₂ : DFA α σ') (s : σ × σ') :
    s ∈ (M₁.inter M₂).accept ↔ s.fst ∈ M₁.accept ∧ s.snd ∈ M₂.accept := by
```

I will rename it to `mem_accept_inter`. My question is should I:
- Make `s` implicit?
- Make all three of `M₁`, `M₂`, `s` implicit?

I think the first option makes sense (it'd mirror the signature of `mem_accept_compl`).

YaelDillies replied:
Iffs should in general have as many implicit arguments as possible so that one can write `iff_lemma.1`/`iff_lemma.2` to get access to the forward and backward implications

meithecatte commented:
Relevant Zulip discussion: [#mathlib4 > Regular languages: the review queue](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/Regular.20languages.3A.20the.20review.20queue/with/514669532)

maemre committed:
Merge branch 'master' into dfa_set_closure

maemre committed:
Computability/DFA: Fix comment spacing.

maemre committed:
Computability/DFA: Better names and API for DFA complement theorems.

maemre committed:
Computability: Add closure properties in terms of Language.IsRegular
